### DISCUSSION TITLE
normalize + aes()?

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#25] [Q&A]
normalize + aes()

**Comment by EvaMaeRey** [comment]
@clauswilke

**Reply by clauswilke** [reply]
?

**Reply by EvaMaeRey** [reply]
An idea for your 'Effective data visualization with ggplot2' workshop.  I still think about + aes(): it's the best kept secret ('advanced tricks and tips') in ggplot2.  Would be helpful to many, I believe, if the syntax were normalized and mainstream, not something people accidentally stumble across (and find to be super game changing).  I think the argument has been made that showing two syntax options would be confusing.  But there's a lot of different way of getting things done in other ggplot2 realms which are 'okay', and people are smart.

**Reply by clauswilke** [reply]
Ok. I can mention it.

**Reply by EvaMaeRey** [reply]
Hooray!

**Reply by EvaMaeRey** [reply]
Hope you have a great workshop!  Wish I could attend!

**Reply by EvaMaeRey** [reply]
@teunbrand are you normalizing this this year too? 🙃  I guess I'm a little late here.

**Reply by teunbrand** [reply]
I think I've used it in a few places, but mostly for reasons of keeping line length shorter rather than other perceived merits.

**Reply by EvaMaeRey** [reply]
'perceived merits' 👌😄

**Reply by aphalo** [reply]
In 'ggspectra' I have `ggplot()` especializations that automatically set the mappings to _x_ and _y_ aesthetics, and the `+ aes()` syntax allows users to add additional mappings.

**Reply by EvaMaeRey** [reply]
Sounds interesting, @aphalo... Is like a case where ggspectra() makes a pretty complete plot and then something like + aes(alpha= extravar) allows for optional variable mapping?  Just a guess, before getting a chance to take a look - hopefully in the next few days!

**Reply by aphalo** [reply]
@EvaMaeRey Yes, your guess is correct. 'ggspectra' is the package, but it defines specializations for `ggplot()` and `autoplot()` methods that have a default argument for `mapping` matching the class of the object passed as argument to `data`. These methods are dispatched, based on the generics defined in 'ggplot2'. Dispatch follows the class of the object passed as argument to `data`.

**Comment by EvaMaeRey** [comment]
I like this discussion about ggplot2's aesthetic mapping's centrality from Winston's book.  Maybe + aes() better reflects this centrality.  Kind of thinking about this leading up to {tidyplots}.  

> In a data graphic, there is a mapping (or correspondence) from properties of the data to visual properties in the graphic. The data properties are typically numerical or categorical values, while the visual properties include the x and y positions of points, colors of lines, heights of bars, and so on. A data visualization that didn’t map the data to visual properties wouldn’t be a data visualization. On the surface, representing a number with an x coordinate may seem very different from representing a number with a color of a point, but at an abstract level, they are the same. Everyone who has made data graphics has at least an implicit understanding of this. For most of us, that’s where our understanding remains.

> **In the grammar of graphics, this deep similarity is not just recognized, but made central.** In R’s base graphics functions, each mapping of data properties to visual properties is its own special case, and changing the mappings may require restructuring your data, issuing completely different plotting commands, or both.

**Comment by teunbrand** [comment]
Why do you like this @EvaMaeRey?
I've seen you use this a bunch in material you've prepared but I've never quite understood why this would be preferable over the convertional `ggplot(mapping)` or `layer(mapping)` ways.

**Reply by EvaMaeRey** [reply]
I think it's a better syntactic representation of how people 'speak their plots into existence' and better 'maps' to the three major steps of specifying a plot (enough to learn about the data - not pub ready necessarily) 🙃: 1) define the data 2) define which visual channels will be used to represent variables in the data and 3) define the mark on the page that is going to take on those visual characteristics. 

The unbundled syntax + aes(all the mappings) is probably easier for newcomers and probably helps with readability which is more to tidyplots concerns, though I don't think this has been studied.  And for folks that grew up on ggplot(data, aes()), it might be less easy to read because at this point it looks weird/wrong/yucky.  🤷‍♀️

For me, it's not just about +aes(all the mappings), but the + aes(some mappings) + layer() + aes(some more variable exploration).   This means you can capture conversations with data (analytic process) astoundingly well (syntax in order of exploration moves) - I think this is kind of awesome and fun!

**Reply by teunbrand** [reply]
Interesting, I hadn't thought about the 3 steps. Now that you're mentioning this, using `+ aes()` also elevates the mapping to the same level as other plot components that also can be `+`'ed.  In terms of this typical figure:

![](https://ggplot2.tidyverse.org/articles/ggplot2_files/figure-html/overview_graphic-1.png)

It isn't wedged in with the data in `ggplot()` and acts as its own component.

**Reply by EvaMaeRey** [reply]
Yes - it definitely matches these figures! And since the data is orthogonal to the variable mappings (like you can have multiple datasets at work in different layers - but with the same declared variable names and mappings at work) it kind of seems like it's uncomfortably 'wedged in with the data'.

**Reply by EvaMaeRey** [reply]
Just thought it interesting that a discussion point in the tidyplots meeting and in 'meet the speaker' was that aes() syntax is part of the reason people create and use non-base ggplot2 syntax frameworks like tidyplots, ggformula, ggblanket.  (and maybe due to how it is typically taught, with the declaration and feeling nested - multiple parenthases)

Maybe part of the trouble with 'aesthetic mapping' is that it is *so* intuitive - so people don't talk this way:  So formal description -- 'we'll map population to the size aesthetic' -- feels like overkill.   Instead we're more likely to say 'let's put weight on the x axis'.  'Color is sex.'  'y is income' and that doesn't feel like a misrepresentation of what's going on.

**Comment by EvaMaeRey** [comment]
Taking aesthetic mapping's centrality a step further, I thought a good blog post title be, 'Aesthetic mapping *is* data visualization'.  

Here's a couple of AI generated blogposts - both of which sound very AI.

ChatGPT: https://chatgpt.com/share/682e41c2-48ec-8005-8460-5f58ec13adce
Claude: https://claude.ai/share/4c87ea51-6063-4e3d-9ac8-0694b1bd08f6

I think I prefer ChatGPT's version but I'm too old-school/pro-school to be okay with 'by [your name here]'.

**Reply by EvaMaeRey** [reply]
The main lesson here, I think, is that writing such a blog post, even if earnestly, is not likely to do much for the + aes() cause.

**Comment by EvaMaeRey** [comment]
Normalizing!

 https://www.linkedin.com/posts/cedscherer_ggplot2-ggplot2-dataviz-activity-7335690317754474496-PioM?utm_source=share&utm_medium=member_desktop&rcm=ACoAAAIFK0wBe0Q89iD-ZEWren9Fjl9ugLy3zB0

![image](https://github.com/user-attachments/assets/0c0544ab-2a93-4232-9e46-203d432a2f45)
